### PR TITLE
collation_fn argument for batched + DBCache source_ fn

### PR DESCRIPTION
**Pull request by kdonbekci**
Hey Tom! 

I've been using `webdataset` for managing our audio data pipeline and ran into a couple of blockers for our use case. 

## user-specified collation function
The first change is regarding the `.batched` function in the `Shorthands` class. In master, you cannot specify a collation function to use akin to what `torch.utils.data.DataLoader` allows. The actual function responsible with batching, `iterators.batched` allows for a user-specified collation function but this is not pushed into the `Shorthands` class. A couple of reasons why this is a good feature to have are allowing batched dictionaries, complicated padding strategies, in-batch ordering of elements, ...

## DBCache missing `source_` attr
The second change is adding the `.source_` function to the `DBCache` class. I was simply following the `readme.ipynb` and could not get the section with the sample caching work. By adding the changes, I can get the expected behavior of the class.

### CODE REVIEW

**Comment by tmbdev** [comment]
Thanks for the patch.
